src/iop/borders.c

```diff
@@ -373,24 +430,55 @@ void modify_roi_out(struct dt_iop_module_t *self, struct dt_dev_pixelpipe_iop_t
   const float size = fabsf(d->size);
   if(size == 0) return;
 
-  if(d->aspect == DT_IOP_BORDERS_ASPECT_CONSTANT_VALUE)
+  const gboolean is_constant_border = d->aspect == DT_IOP_BORDERS_ASPECT_CONSTANT_VALUE;
+
+  dt_iop_basis_t basis = d->basis;
+  if (is_constant_border && !d->max_border_size)
+  {
+    // the legacy behaviour is, when a constant border is used and the
+    // max_border_size flag is set, the width is always used as basis.
+    basis = DT_IOP_BORDERS_BASIS_WIDTH;
+  }
+  else if (basis == DT_IOP_BORDERS_BASIS_AUTO)
   {
+    // automatic/legacy/default behaviour:
     // for a constant border be sure to base the computation on the
-    // larger border, failing that the border will have a difference
+    // larger border, failing that the border will have a different
     // size depending on the orientation.
+    // for all other borders use the width.
+    basis = is_constant_border ? DT_IOP_BORDERS_BASIS_LONGEST : DT_IOP_BORDERS_BASIS_WIDTH;
+  }
+  if(basis == DT_IOP_BORDERS_BASIS_LONGEST)
+  {
+    basis = roi_in->width > roi_in->height ? DT_IOP_BORDERS_BASIS_WIDTH : DT_IOP_BORDERS_BASIS_HEIGHT;
+  }
+  else if(basis == DT_IOP_BORDERS_BASIS_SHORTEST)
+  {
+    basis = roi_in->width < roi_in->height ? DT_IOP_BORDERS_BASIS_WIDTH : DT_IOP_BORDERS_BASIS_HEIGHT;
+  }
 
-    if(roi_in->width > roi_in->height || !d->max_border_size)
-    {
-      // this means: relative to width and constant for height as well:
-      roi_out->width = roundf((float)roi_in->width / (1.0f - size));
-      roi_out->height = roi_in->height + roi_out->width - roi_in->width;
-    }
-    else
-    {
-      // this means: relative to height and constant for width as well:
-      roi_out->height = roundf((float)roi_in->height / (1.0f - size));
-      roi_out->width = roi_in->width + roi_out->height - roi_in->height;
-    }
+  assert(basis == DT_IOP_BORDERS_BASIS_WIDTH || basis == DT_IOP_BORDERS_BASIS_HEIGHT);
+
+  const int *basis_in = NULL, *other_in = NULL;
+  int *basis_out = NULL, *other_out = NULL;
+  
+  #define DT_IOP_BORDERS_ASSIGN(b, b_uc, o) \
+    basis_in = &roi_in->b, basis_out = &roi_out->b, \
+    other_in = &roi_in->o, other_out = &roi_out->o, \
+    basis = DT_IOP_BORDERS_BASIS_ ## b_uc
+  #define DT_IOP_BORDERS_ASSIGN_width DT_IOP_BORDERS_ASSIGN(width, WIDTH, height)
+  #define DT_IOP_BORDERS_ASSIGN_height DT_IOP_BORDERS_ASSIGN(height, HEIGHT, width)
+  #define DT_IOP_BORDERS_ASSIGN_BASIS(basis) DT_IOP_BORDERS_ASSIGN_ ## basis
+
+  if(basis == DT_IOP_BORDERS_BASIS_WIDTH)
+    DT_IOP_BORDERS_ASSIGN_BASIS(width);
```

Comment:
Wow! I'm guilty of using some funky macros myself, but this breaks my head early in the morning! ;-)
You're  introducing 7 lines of nested macros to save 8 lines of code... I'm usually aiming for a better payoff, but stones/glass houses so I won't say anything! ;-)

Thanks for implementing this feature! Open source thrives when people pitch in so scratch their own itches. Hope you find many other things that annoy you!

Reply:
Yeah, I know, haha. Otherwise it would be a lot of repetition though, like this at least the code below the macros and the intention behind it is more clear. I didn't feel like making a data structure and some functions just for this, which would add runtime overhead, or completely redesigning the code. It works for now!

Thank you for the feedback!